source/core/Events.cpp

```diff
@@ -1048,6 +1048,9 @@ VIREO_FUNCTION_SIGNATUREV(WaitForEventsAndDispatch, WaitForEventsParamBlock)
         Int32 &staticCount = eventInfo->eventStructInfo[eventStructIndex].setCount;
 
         UInt32 msTimeout = timeOutPtr ? *timeOutPtr : -1;
+        if (msTimeout == 0) {
+            return arguments->branchTarget;
```

Comment:
This isn't correct, you're not executing any of the event dispatch code below, so if an event is actually available, you will not dispatch to it and handle it.  Your tests only test events with a single timeout case so they do not fully test this.
Carlos's suggestion may actually be the correct fix.

Reply:
Took Carlos's suggestion. And also have updated the tests.